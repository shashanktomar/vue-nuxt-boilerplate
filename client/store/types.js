// @flow

export const SET_COMMENTS = 'root/SET_COMMENTS'
export const UPDATE_RATING = 'root/UPDATE_RATING'
