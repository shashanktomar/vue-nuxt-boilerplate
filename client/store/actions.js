// @flow

import comments from '~/logics/api/comments'
import * as mutationTypes from '~/store/types'

import type { ActionContext } from 'vuex'
import type { StateType } from '~/types'

const actions = {
  async fetchComments({ commit, _state }: ActionContext<StateType>) {
    const commentsList = await comments.fetchComments(this.$axios)
    commit(mutationTypes.SET_COMMENTS, commentsList)

    // Uncomment next line to see typing in action:
    // console.log(_state.comments, _state.fake)

    return commentsList
  },
}

export default actions
