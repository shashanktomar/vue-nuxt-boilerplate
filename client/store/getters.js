// @flow

import type { StateType } from '~/types'

const getters = {
  hasComments (state: StateType): boolean {
    return Boolean(state.comments && state.comments.length > 0)
  },
}

export default getters
