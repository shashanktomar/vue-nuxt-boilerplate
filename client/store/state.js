// @flow

import type { StateType } from '~/types'

function createState (): StateType {
  return {
    'comments': null,
  }
}

export default createState
