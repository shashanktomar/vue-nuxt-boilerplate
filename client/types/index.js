// @flow

export type RawCommentType = {
  id: number,
  body: string,
  email: string
}

// We add rating only on the client (for demo purposes)
export type CommentType = RawCommentType & {
  rating: number
}

export type CommentPayloadType = {
  commentId: number,
  delta: number
}

export type StateType = {
  comments: ?Array<CommentType>
}
