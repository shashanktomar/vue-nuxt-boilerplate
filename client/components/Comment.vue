<template>
  <div :class="computedClasses">
    <p :class="$style.author">{{ comment.email }}</p>

    <p :class="$style.body">{{ comment.body }}</p>

    <div :class="$style.rating">
      <button
        :class="$style.ratingUp"
        @click="changeRating(comment.id, 1)"
      >
        +
      </button>

      <span :class="$style.number">{{ comment.rating }}</span>

      <button
        :class="$style.ratingDown"
        @click="changeRating(comment.id, -1)"
      >
        -
      </button>
    </div>
  </div>
</template>

<script>
// @flow

import Vue from 'vue'
import Component from 'nuxt-class-component'
import { Prop } from 'vue-property-decorator'
import { Mutation } from 'vuex-class'

import * as mutationTypes from '~/store/types'
import type { CommentType, CommentPayloadType } from '~/types'

// @vue/component
@Component()
/**
 * Comment component is used to represent a single user's comment.
 */
export default class Comment extends Vue {
  @Mutation(mutationTypes.UPDATE_RATING)
  /**
   * This is a wrapped mutation from the vuex.
   */
  updateRating: (CommentPayloadType) => void

  @Prop()
  /**
   * Passed comment from the parent component.
   */
  comment: CommentType

  /**
   * Changes comment's rating.
   * Can be used to increase or decrease comment's rating.
   *
   * @param commentId - Comment's identifier to change rating.
   * @param delta - Delta value to change rating value.
   */
  changeRating (commentId: number, delta: number) {
    // Uncomment next line to test typing:
    // console.log(this.comment.missingKey)
    this.updateRating({ commentId, delta })
  }

  /**
   * Defines which color borders are.
   * Uses `comment`'s rating to choose color.
   *
   * @returns Pairs of class names and boolean values if they should be applied.
   */
  get computedClasses () {
    return {
      [this.$style.commentComponent]: true,
      [this.$style.commentPositive]: this.comment.rating > 0,
      [this.$style.commentNegative]: this.comment.rating < 0,
    }
  }
}
</script>

<style lang="scss" module>
@import '~/scss/variables';

.commentComponent {
  background: $color-white;
  max-width: 15.6rem;
  border: 0.12rem solid;
  padding: 0.74rem;
  margin: 0.5rem;
  position: relative;
}

.commentPositive {
  border-color: $button-color-green;
}

.commentNegative {
  border-color: $button-color-red;
}

.author {
  padding-bottom: 0.74rem;
  font-weight: 700;
  border-bottom: 0.12rem solid;
}

.body {
  text-align: justify;
}

.rating {
  position: absolute;
  bottom: 0.3rem;
  left: 50%;
  transform: translateX(-50%);

  .number {
    // This will be available as `this.$style.number`
    font-weight: 700;
  }

  button {
    margin: 0 0.5rem;
    border: none;
    outline: none;
    cursor: pointer;
    width: 2rem;
  }
}

.ratingUp {
  background: $button-color-green;
}

.ratingDown {
  background: $button-color-red;
}
</style>
