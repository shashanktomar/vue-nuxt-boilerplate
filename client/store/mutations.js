// @flow

import * as mutationTypes from '~/store/types'

import type {
  CommentType,
  CommentPayloadType,
  StateType,
  RawCommentType,
} from '~/types'

const mutations = {
  [mutationTypes.SET_COMMENTS]: (
    state: StateType,
    comments: Array<RawCommentType>,
  ) => {
    const updatedComments: Array<CommentType> = []

    for (const comment of comments.slice(0, 10)) {
      const newOne = {
        ...comment,
        'rating': 0, // try to comment out this line
      }
      updatedComments.push(newOne)
    }

    state.comments = updatedComments
  },

  [mutationTypes.UPDATE_RATING]: (state: StateType, {
    commentId,
    delta,
  }: CommentPayloadType) => {
    if (!state.comments) return

    const commentIndex = state.comments.findIndex((comment: CommentType) => {
      return comment.id === commentId
    })

    if (!state.comments || !state.comments[commentIndex]) return

    state.comments[commentIndex].rating += delta
  },
}

export default mutations
