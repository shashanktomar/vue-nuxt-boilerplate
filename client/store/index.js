// @flow

import Vuex from 'vuex'
import createState from '~/store/state'
import mutations from '~/store/mutations'
import getters from '~/store/getters'
import actions from '~/store/actions'

const createStore = () => {
  return new Vuex.Store({
    strict: process.env.NODE_ENV !== 'production',
    state: createState(),
    getters,
    actions,
    mutations,
  })
}

export default createStore
