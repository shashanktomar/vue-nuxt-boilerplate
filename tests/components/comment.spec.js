// @flow

import { mount, createLocalVue } from '@vue/test-utils'

import Comment from '~/components/Comment'
import * as types from '~/store/types'

import { storeFactory, commentFactory } from '../fixtures/vuex'

const localVue = createLocalVue()

describe('unit tests for Comment component', () => {
  let comment
  let store

  beforeEach(() => {
    comment = commentFactory.build({ rating: 0 })
    store = storeFactory.build({ state: { comments: [comment] } }, { localVue })
  })

  // Uncomment this line to test typing:
  // test(12, {})

  test('should have two buttons', () => {
    const wrapper = mount(Comment, {
      store,
      localVue,
      propsData: { comment },
    })

    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.is(Comment)).toBeTruthy()
    expect(wrapper.findAll('button')).toHaveLength(2)
  })

  test('should have correct values', () => {
    const wrapper = mount(Comment, {
      store,
      localVue,
      propsData: { comment },
    })

    expect(
      wrapper
        .find('.body')
        .text()
        .trim()
    ).toEqual(comment.body)
    expect(
      wrapper
        .find('.author')
        .text()
        .trim()
    ).toEqual(comment.email)
    expect(
      wrapper
        .find('.number')
        .text()
        .trim()
    ).toEqual(comment.rating.toString())
  })

  test('should increment rating', () => {
    const delta = 1
    const newRating = comment.rating + delta
    const wrapper = mount(Comment, {
      store,
      localVue,
      propsData: { comment },
    })

    wrapper.vm.$store.commit(types.UPDATE_RATING, {
      commentId: comment.id,
      delta,
    })
    expect(wrapper.vm.$store.state.comments[0].rating).toEqual(newRating)

    expect(wrapper.props().comment.rating).toEqual(newRating)
    expect(wrapper.classes()).toContain(wrapper.vm.$style.commentPositive)
  })

  test('should decrement rating', async () => {
    const delta = -2
    const newRating = comment.rating + delta
    const wrapper = mount(Comment, {
      store,
      localVue,
      propsData: { comment },
    })

    wrapper.vm.$store.commit(types.UPDATE_RATING, {
      commentId: comment.id,
      delta,
    })
    expect(wrapper.vm.$store.state.comments[0].rating).toEqual(newRating)

    expect(wrapper.props().comment.rating).toEqual(newRating)
    expect(wrapper.classes()).toContain(wrapper.vm.$style.commentNegative)
  })
})

describe('snapshot tests for Comment component', () => {
  let comment
  let store

  beforeAll(() => {
    // We need a seed here to be consistent for snapshot testing:
    comment = commentFactory.build({}, { seed: 8874 })
    store = storeFactory.build({ state: { comments: [comment] } }, { localVue })
  })

  test('should match the snapshot', () => {
    const wrapper = mount(Comment, {
      store,
      localVue,
      propsData: { comment },
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
