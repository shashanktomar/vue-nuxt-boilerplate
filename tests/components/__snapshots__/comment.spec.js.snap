// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests for Comment component should match the snapshot 1`] = `
"<div class=\\"commentComponent commentPositive\\"><p class=\\"author\\">Mellie_Gibson59@hotmail.com</p> <p class=\\"body\\">Sequi provident inventore dicta. Quia explicabo assumenda eum.</p> <div class=\\"rating\\"><button class=\\"ratingUp\\">
      +
    </button> <span class=\\"number\\">9</span> <button class=\\"ratingDown\\">
      -
    </button></div></div>"
`;
